### ISSUE TITLE
Stop using `Base.@irrational`

### ISSUE THREAD

**Issue by hyrodium** [JuliaMath/IrrationalConstants.jl#18]
This macro should not be used outside of `Base`. See https://github.com/JuliaLang/julia/pull/47103 and https://github.com/JuliaLang/julia/issues/46531.

**Comment by devmotion** [comment]
I'm well aware of this issue (see, e.g., https://github.com/JuliaMath/IrrationalConstants.jl/pull/12#issuecomment-1003453724 and https://github.com/TuringLang/MCMCDiagnosticTools.jl/pull/63#discussion_r1072776711). The main/only point of this package is that already existing definitions of irrationals, floating around in the ecosystem, are defined in a common package so it's a lightweight dependency and thereby makes it less likely that two packages define the same constants. Most constants have been defined in StatsFuns for a long time before this package was created and have been used in particular in different stats packages; IIRC then at some point someone wanted to use them in SpecialFunctions which does not want to add a dependency on StatsFuns (and StatsFuns even depends on SpecialFunctions). Of course, it's only the second best solution but I think it's definitely better than every package trying to redefine irrational constants.

**Comment by hyrodium** [comment]
I see. Sorry for bothering you.

There's one more thing I'd like to ask you about. Is there any problems if we define the constants with `IrrationalConstant{sym} <: AbstractIrrational` and `@irrationalconstant` (a macro to define `IrrationalConstant` just like `Base.@irrational`)? I thought this solution can resolve the type piracy and does not produce any other problems.

**Comment by devmotion** [comment]
Yes, I assume the way to fix this are custom types that are not owned by base, as suggested in https://github.com/JuliaLang/julia/issues/46531#issue-1353977320. I'm not sure how many methods in base are defined for `Irrational` explicitly - the annoying thing with custom types would be that we would miss functionality that does not support `AbstractIrrational`.

**Comment by devmotion** [comment]
But at least a quick search does not reveal anything: https://github.com/JuliaLang/julia/search?p=1&q=Irrational

**Comment by hyrodium** [comment]
I have not checked all of the code, but it seems we need to define these methods at least.
```julia
show(io::IO, x::IrrationalConstant{sym}) where {sym} = print(io, sym)

function show(io::IO, ::MIME"text/plain", x::IrrationalConstant{sym}) where {sym}
    if get(io, :compact, false)::Bool
        print(io, sym)
    else
        print(io, sym, " = ", string(float(x))[1:min(end,15)], "...")
    end
end

==(::IrrationalConstant{s}, ::IrrationalConstant{s}) where {s} = true
<(::IrrationalConstant{s}, ::IrrationalConstant{s}) where {s} = false
<=(::IrrationalConstant{s}, ::IrrationalConstant{s}) where {s} = true
hash(x::IrrationalConstant, h::UInt) = 3*objectid(x) - h  # I'm not much familiar with `hash`. just copied from Base.
```

**Comment by devmotion** [comment]
I think that we would want to avoid the design of `Irrational` since that leads to exactly the same problems if some tries to extend our irrational type in a package. Basing irrationals on a type parameter seems flawed. Instead we would want separate singleton types for each irrational constants, as proposed in the issue linked above.

If you don't beat me to it, I might have time to look into it later this evening.

**Comment by devmotion** [comment]
Fixed by #19, will be released as IrrationalConstants 0.2.
